### PR TITLE
Add support for shader options within PCF filtering for directional light shadows

### PR DESCRIPTION
**Pull request by moudgils**
## What does this PR do?

This PR adds support for shader options for different PCF filtering modes. This will allow un-used filtering modes to be baked out and help with register usage for forward pass. We will need to build our variants for o_directional_shadow_filtering_sample_count in order to enable this optimization. The shader option support was only added for directional lights for this PR. Projected lights are still using ViewSrg to extract sample count

## How was this PR tested?

Tested Editor (DX12 and Vk) by adding directional light as well as projected light and ensuring that the PCF sample count still works as desired.

### CODE REVIEW

**Review by AMZN-alexpete — APPROVED** [review]
Nbd but I kinda wish we stuck with naming stuff either sample count or sample count mode instead of a mix.

**Comment by moudgils** [comment]
> Nbd but I kinda wish we stuck with naming stuff either sample count or sample count mode instead of a mix.

Sample count can be any number between 0-64 but sample count mode can only be 0-4. The main reason for it is that when you have a shader option and want to build out variants you can build variants for exactly 0-4 as oppose to having to build variants for 0-64 and exploding the variant tree. Changing everything (i.e Editor) to sample count mode is probably the better approach here but is a bigger task.

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/CoreLights/ShadowConstants.h [hunk 1]
```diff
@@ -27,7 +27,8 @@ namespace AZ
         static constexpr ShadowmapSize MinShadowmapImageSize = ShadowmapSize::Size256;
         static constexpr ShadowmapSize MaxShadowmapImageSize = ShadowmapSize::Size2048;
 
-        // This matchs ShadowFilterMethod in CoreLights/ViewSrg.azsli
+        // This matches ShadowFilterMethod in DirectionalLightShadowCalculator.azsli
+        // and represents m_shadowFilterMethod in CoreLights/ViewSrg.azsli
         enum class ShadowFilterMethod : uint32_t
```
Comment (AMZN-Gene) [inline Gems/Atom/Feature/Common/Code/Include/Atom/Feature/CoreLights/ShadowConstants.h:32]: Why does this enum have one value, but the shader allows for 4 (None, Pcf, Esm, and EsmPcf)?

File: Gems/Atom/Feature/Common/Code/Source/CoreLights/DirectionalLightFeatureProcessor.cpp [hunk 2]
```diff
@@ -585,9 +590,25 @@ namespace AZ
                 AZ_Warning(FeatureProcessorName, false, "Sampling count exceed the limit.");
                 count = Shadow::MaxPcfSamplingCount;
             }
+
+            //Remap the count value to an enum value associated with that count.
+            ShadowFilterSampleCount samplingCountMode = ShadowFilterSampleCount::Tap0;
```
Comment (AMZN-Gene) [inline Gems/Atom/Feature/Common/Code/Source/CoreLights/DirectionalLightFeatureProcessor.cpp]: just making sure, _samplingCountMode_ will never be _ShadowFilterSampleCount::Tap0_?

**Review by AMZN-Gene — APPROVED** [review]

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/CoreLights/ShadowConstants.h (hunk 1 above)
Comment (moudgils) [reply Gems/Atom/Feature/Common/Code/Include/Atom/Feature/CoreLights/ShadowConstants.h:32]: Maybe there is a misunderstanding here but this enum has four values which should match the one in the shader

File: Gems/Atom/Feature/Common/Code/Source/CoreLights/DirectionalLightFeatureProcessor.cpp (hunk 2 above)
Comment (moudgils) [reply Gems/Atom/Feature/Common/Code/Source/CoreLights/DirectionalLightFeatureProcessor.cpp]: yeah. Not at the moment. We could in theory add support for that mode in future for low end devices.

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/CoreLights/ShadowConstants.h [hunk 3]
```diff
@@ -38,6 +39,18 @@ namespace AZ
             Count
         };
 
+        // This matches ShadowFilterSampleCount in DirectionalLightShadowCalculator.azsli
+        // and represents m_filteringSampleCountMode in CoreLights/ViewSrg.azsli
+        enum class ShadowFilterSampleCount : uint32_t
+        {
+            Tap0 = 0,
```
Comment (VickyAtAZ) [inline Gems/Atom/Feature/Common/Code/Include/Atom/Feature/CoreLights/ShadowConstants.h]: Tap0 seems useless and can be removed.

File: Gems/Atom/Feature/Common/Code/Source/CoreLights/DirectionalLightFeatureProcessor.cpp (hunk 2 above)
Comment (VickyAtAZ) [inline Gems/Atom/Feature/Common/Code/Source/CoreLights/DirectionalLightFeatureProcessor.cpp]: You can set it to Tap16 and get rid of tap0 everywhere

**Review by VickyAtAZ — APPROVED** [review]

**Pull request merged**